### PR TITLE
Possibility to create Google compute client

### PR DESCRIPTION
**Pull request by mkasztelnik**
TODOs:
  
  * [x] Create google compute client using for library
  * [x] Get flavors (`FlavorManager`)
  * [x] Get flavor data (`FlavorManager`)
  * [x] Get vms (`VmMonitoringWorker`)
  * [x] Get vm data (`VmUpdater`)
  * [x] Create tags for vm (`VmTagsCreator`)
  * [x] Get templates	 (`VmTemplateMonitoringWorker`)
  * [x] Destroy vm (`VirtualMachine`)
  * [x] Managing vm (`VirtualMachine`)
  * [ ] Save template	(`VirtualMachineTemplate`)
  * [x] Destroy template (`VirtualMachineTemplate`)
  * [x] Create vm (`VmCreator`)
  * [x] Import key pair (`UserKey`)
  * [x] Delete key pair (`UserKey`)
  * [x] Get template data (`VmtUpdater`)
  * [x] Permit all for http/https/tcp/udp
  * [x] Initial configuration

### CODE REVIEW

**Comment by mkasztelnik** [comment]
Example compute site configuration: 

```json
{"provider":"Google","google_key_location":"p12_path","google_key_string":"xxx","google_project":"project_id","google_client_email":"xxx@developer.gserviceaccount.com"}
```

**Comment by mkasztelnik** [comment]
@nowakowski it seems like google can charge for every 10 minutes of VM usage: https://cloud.google.com/compute/pricing#machinetype - what do you think about it? Should we stick to 1h charging interval or add some extension to support 10 minutes interval?

**Comment by mkasztelnik** [comment]
@nowakowski it is even better, after 10 minutes you are charged in 1 minute intervals...

**Comment by nowakowski** [comment]
Hm, this is kind of bad - charge interval is a config variable and it can be changed to 10 minutes without any problems but it applies to all of Atmosphere and not to any specific compute site. :( Keep 1 hour for now.

**Comment by mkasztelnik** [comment]
to create new server:

```ruby
disk_hsh = {
   :name => "disk name can be the same as server name",
   :size_gb => 10, # smallest possible value
   :zone_name => "us-central1-a",
   :source_image => "debian-7-wheezy-v20140408"
}

disk = client.disks.create(disk_hsh)
disk.wait_for { disk.ready? }
disks = [disk]

server_hsh = {
 :name=>"mktest",
 :machine_type=>"n1-standard-1",
 :zone_name=>"us-central1-a",
 :username=>"marek",
 :public_key_path=>"/home/marek/.ssh/id_rsa.pub",
 :disks=> disks
}

server = client.create(server_hsh)
```

There is also possible to set that disk should be deleted while server is terminated - need to check how to set it.

**Comment by mkasztelnik** [comment]
Creating image from instance:

* Instance **need** to be deleted before creating new image. Disk **cannot** be attached to any instance

```ruby
disk = ... # disk fog object
client = ... # fog google compute client

disk_uri = "#{disk.zone}/disks/#{disk.name}"
img = client.insert_image('name', description: 'description', sourceDisk: disk_uri)
```

* At the end we can start new instance, but it will receive different ip address.

**Comment by nowakowski** [comment]
@mkasztelnik Could you be a bit more specific? We're not intimately familir with the Google Compute cloud management model. In particular, I would like to ask the following questions:
  - Does your first bullet list the conditions which need to be satisfied in order for the image creation request to succeed?
  - If the above is true, how should one interpret the notion that the "instance needs to be deleted"? Do you mean suspended?
  - If the instance must be physically destroyed to feed the image creation process, what happens if the image cannot be successfully created? Can the instance be recovered?
  - Does the GC virtualization model entail storage of all instance data (OS + userspace) on a single disk image which is preserved in the cloud, or is the instance assembled from a separate OS image and userspace data like e.g. in Amazon EC2?
  - What is a "disk zone"?

In general, if the creation of an image requires destruction of an instance (which seems relly weird), I would suggest asking DH to provide an appropriate warning popup ("molly guard") whenever someone attempts to save a ApplianceType based on an existing devmode instance which is backended by a VM which runs on a cloud site of type 'google' - and the proceeding with the instance deletion/image creation request regardless.

**Comment by mkasztelnik** [comment]
Simply google compute has similar solution as Azure. First you are creating disk with OS, next you "attache" CPU and RAM into the disk. Google compute understand instance deleting by removing CPU and RAM from the machine. Persistent storage with OS data are stored on disk, which you need to manage separately.

So to save Appliance as Appliance Set in our case we need to delete Virtual Machine without destroying disk attached to this Virtual Machine. Only then we can create image (aka. atmosphere Virtual Machine Template). The problem is connected with the way how atmosphere works. It assumes that VM attached into appliance always exists (even when it is in saving mode). This is not true in Google compute case.

**Comment by mkasztelnik** [comment]
@nowakowski, @tbartynski, @Nuanda, @bwilk please read last 3 comments and if you have proposition how it can be solved in atmosphere (not in GUI - we will deal with it later on) please leave a comment :smile:

**Comment by mkasztelnik** [comment]
I propose to make this PR as ready for review, make the review process merge it and then create 2 additional PRs: first one with "save as" capability and the second one with "save" capability.

@dice-cyfronet/atmo-dev-team please take a look at it.

**Comment by mkasztelnik** [comment]
@dice-cyfronet/atmo-dev-team  unfortunately we need to make it very fast, because my free google compute grant finishes after 7 days - so please treat this PR as priority task on your TODO list.

File: config/initializers/unify_google_compute.rb [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        end
+
+        private
+
+        def create_disk(params)
+          disk_params = {
+            name: params[:name],
+            # think about taking it from flavor or user requirement.
+            # Flavor disk is currently set to 0 so this also need to be
+            # rethinked.
```
Comment (nowakowski) [inline config/initializers/unify_google_compute.rb]: rethought ;)

File: app/workers/atmosphere/vm_template_monitoring_worker.rb [hunk 2]
```diff
@@ -9,9 +9,17 @@ def perform(site_id)
...
+        filters = site.template_filters ? JSON.parse(site.template_filters) : {}
+
         logger.debug { "#{jid}: getting images state for site #{site_id} from compute site" }
-        update_images(site, site.cloud_client.images.all(filters))
+
+        images = if filters.empty?
+                   site.cloud_client.images.all
+                 else
+                   site.cloud_client.images.all(filters)
+                 end
```
Comment (Nuanda) [inline app/workers/atmosphere/vm_template_monitoring_worker.rb]: Why can't we call images(filters) when filters == {} anymore?

Comment (mkasztelnik) [inline app/workers/atmosphere/vm_template_monitoring_worker.rb]: Because Google compute `all` method does not allow any parameter.

File: config/initializers/unify_google_compute.rb [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+        alias_method :status_orig, :status
+        def status
+          s = status_orig
+
+          if s == 'READY'
+            'active'
+          else
+            s
+          end
```
Comment (Nuanda) [inline config/initializers/unify_google_compute.rb]: maybe: `status_orig == 'READY' ? 'active' : status_orig`

**Comment by Nuanda** [comment]
:+1:

File: app/controllers/atmosphere/api/v1/appliance_types_controller.rb [hunk 4]
```diff
@@ -93,6 +78,14 @@ def endpoint_payload
 
         private
 
+        def check_for_conflict!(appl)
+          vm = appl.virtual_machines.first
+          if vm && vm.state.saving?
+            raise Atmosphere::Conflict,
+                  'It is not allowed to save application twice'
```
Comment (nowakowski) [inline app/controllers/atmosphere/api/v1/appliance_types_controller.rb]: Może "This service is already being saved. Please wait for this operation to conclude before attempting to save the service again."

File: config/initializers/unify_google_compute.rb [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        def create_disk(params)
+          disk_params = {
+            name: params[:name],
+            # think about taking it from flavor or user requirement.
+            # Flavor disk is currently set to 0 so this also need to be
+            # rethinked.
+            size_gb: 10,
+            zone_name: params[:zone],
+            source_image: params[:image_id],
+            autoDelete: true
```
Comment (nowakowski) [inline config/initializers/unify_google_compute.rb]: Is camel case appropriate here?

File: config/initializers/unify_google_compute.rb [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        end
+
+        def stop
+          raise Atmosphere::UnsupportedException,
+                'Fog des not support google compute stop action'
+        end
+
+        def pause
+          raise Atmosphere::UnsupportedException,
+                'Google compute des not support pause action'
```
Comment (nowakowski) [inline config/initializers/unify_google_compute.rb]: ...does...

File: config/initializers/unify_google_compute.rb [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+          Time.parse(creation_timestamp)
+        end
+
+        def id
+          name
+        end
+
+        def stop
+          raise Atmosphere::UnsupportedException,
+                'Fog des not support google compute stop action'
```
Comment (nowakowski) [inline config/initializers/unify_google_compute.rb]: ...does...

File: config/initializers/unify_google_compute.rb [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        end
+
+        def pause
+          raise Atmosphere::UnsupportedException,
+                'Google compute des not support pause action'
+        end
+
+        def suspend
+          raise Atmosphere::UnsupportedException,
+                'Google compute des not support suspend action'
```
Comment (nowakowski) [inline config/initializers/unify_google_compute.rb]: ...does...

File: config/initializers/unify_google_compute.rb [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+        end
+
+        def start
+          raise Atmosphere::UnsupportedException,
+                'Fog des not support google compute start action'
```
Comment (nowakowski) [inline config/initializers/unify_google_compute.rb]: ...does...

File: app/controllers/atmosphere/api/v1/appliance_types_controller.rb (hunk 4 above)
Comment (mkasztelnik) [inline app/controllers/atmosphere/api/v1/appliance_types_controller.rb]: This is original message, I did not change here anything. If you believe this message should be improved please do it in another PR.

File: config/initializers/unify_google_compute.rb (hunk 5 above)
Comment (mkasztelnik) [inline config/initializers/unify_google_compute.rb]: I need to be camel case because it is converted into json body request send directly into google compute and google compute requires camel case.

File: config/initializers/unify_google_compute.rb (hunk 6 above)
Comment (mkasztelnik) [inline config/initializers/unify_google_compute.rb]: Will be fixed in my next commit :smile:

File: app/controllers/atmosphere/api/v1/appliance_types_controller.rb (hunk 4 above)
Comment (nowakowski) [inline app/controllers/atmosphere/api/v1/appliance_types_controller.rb]: @mkasztelnik Hmm, I've often received PR comments regarding sections of code which were not actually part of the pending PR. Many of those were from you - and I've never made a fuss about it. Let's not overbureaucratize, shall we? ;)

Comment (mkasztelnik) [inline app/controllers/atmosphere/api/v1/appliance_types_controller.rb]: Fixed in https://github.com/dice-cyfronet/atmosphere/pull/138

**Comment by nowakowski** [comment]
:+1:
